R/derive_param_computed.R

```diff
@@ -147,48 +196,246 @@
+      return(dataset)
+    }
+
+    hori_data <- inner_join(hori_data, hori_const_data, by = vars2chr(constant_by_vars))
+  }
+
+  # add analysis value (AVAL) and parameter variables, e.g., PARAMCD
+  hori_data <- hori_data %>%
+    # keep only observations where all analysis values are available
+    filter(!!!parse_exprs(map_chr(
+      analysis_vars_chr,
+      ~ str_c("!is.na(", .x, ")")
+    ))) %>%
+    process_set_values_to(exprs(!!analysis_var := !!analysis_value)) %>%
+    process_set_values_to(set_values_to) %>%
+    select(-all_of(analysis_vars_chr[str_detect(analysis_vars_chr, "\\.")]))
+
+  bind_rows(dataset, hori_data)
+}
+
+#' Asserts `parameters` Argument and Converts to List of Expressions
+#'
+#' The function asserts that the argument is a character vector or a list of
+#' expressions. If it is a character vector, it converts it to a list of
+#' symbols.
+#'
+#' @param parameters The argument to check
+#'
+#' @param optional Is the checked argument optional? If set to `FALSE` and
+#'   `parameters` is `NULL` then an error is thrown.
+#'
+#' @return The `parameters` argument (converted to a list of symbol, if it is a
+#'   character vector)
+#'
+#' @keywords other_advanced
+#' @family other_advanced
+assert_parameters_argument <- function(parameters, optional = TRUE) {
+  assert_logical_scalar(optional)
+  if (optional && is.null(parameters)) {
+    return(invisible(parameters))
+  }
+
+  if (typeof(parameters) == "character") {
+    parameters <- map(parameters, sym)
+  } else {
+    if (!inherits(parameters, "list") || any(!map_lgl(
+      parameters,
+      ~ is_call(.x) || is_expression(.x)
+    ))) {
+      abort(
+        paste0(
+          "`",
+          arg_name(substitute(parameters)),
+          "` must be a character vector or a list of expressions but it is ",
+          what_is_it(parameters),
+          "."
+        )
+      )
+    }
+  }
+  parameters
+}
+
+#' Creating Temporary Parameters and `<variable>.<parameter>` Variables
+#'
+#' The function creates temporary parameters and variables of the form
+#' `<variable>.<parameter>`, e.g., `AVAL.WEIGHT`.
+#'
+#' @param dataset Input dataset
+#'
+#' @param by_vars By variables
+#'
+#' @param parameters List of parameter codes
+#'
+#'   The input dataset is restricted to the specified parameter codes. If an
+#'   expression is specified, a new parameter code is added to the input
+#'   dataset. The name of the element defines the parameter code and the
+#'   expression the observations to select.
+#'
+#'   *Permitted Values:* A character vector of `PARAMCD` values or a list of expressions
+#'
+#' @param analysis_value
+#'
+#'   All variables of the form `<variable>.<parameter>` like `AVAL.WEIGHT` are
+#'   added to the input dataset. They are set to the value of the variable for
+#'   the parameter. E.g., `AVAL.WEIGHT` is set to the value of `AVAL` where
+#'   `PARAMCD == "WEIGHT"`.
+#'
+#'   *Permitted Values:* An unquoted expression
+#'
```

Comment:
should we make this part consistent with `derive_param_computed`? 
The following code is not working
```r
get_hori_data(
  advs,
  by_vars = exprs(USUBJID, VISIT),
  parameters = exprs(SYSBP, DIABP),
  analysis_value = (AVAL.SYSBP + 2 * AVAL.DIABP) / 3,
  filter = NULL
)
```
below one works after `expr` is added for `analysis_value`

```r
get_hori_data(
  advs,
  by_vars = exprs(USUBJID, VISIT),
  parameters = exprs(SYSBP, DIABP),
  analysis_value = expr((AVAL.SYSBP + 2 * AVAL.DIABP) / 3),
  filter = NULL
)
```

Reply:
Yes, good point. I have update `get_hori_data()` such that it handles the arguments in the same way as `derive_param_computed()`.